Gems/ROS2/Code/Include/ROS2/Sensor/Events/EventSourceAdapter.h

```diff
@@ -0,0 +1,190 @@
+        {
+        };
+
+        template<
+            template<
+                template<typename...>
+                class, // EventType
+                template<typename...>
+                class, // EventHandlerType
+                typename...> // Event parameters
+            class Base,
+            template<typename...>
+            class EventType,
+            template<typename...>
+            class EventHandlerType,
+            typename... Args>
+        struct is_specialization_of<Base, Base<EventType, EventHandlerType, Args...>> : AZStd::true_type
+        {
+        };
+    } // namespace Internal
+
+    //! Class adapting event source (ROS2::SensorEventSource) to configurable working frequency. This is handled via adapted event, in
+    //! a similar manner like it is done in SensorEventSource. EventSourceAdapter has its internal handler that connects to
+    //! SensorEventSource source event, and signals adapted event according to frequency set (ROS2::EventSourceAdapter::Configure).
+    //! User can connect to this event using ROS2::EventSourceAdapter::ConnectToAdaptedEvent method. This class should be used, instead
+    //! of using directly a class derived from SensorEventSource, when specific working frequency is required. Following this path, user can
+    //! still use source event - ROS2::EventSourceAdapter::ConnectToSourceEvent. This template has to be resolved using a class derived from
+    //! SensorEventSource.
+    //! @see ROS2::SensorEventSource
+    template<class EventSourceT>
+    class EventSourceAdapter
+    {
+    public:
+        // Require non-abstract type derived from SensorEventSource.
+        static_assert(Internal::is_specialization_of<SensorEventSource, typename EventSourceT::SourceBaseType>::value);
+        static_assert(AZStd::is_base_of<typename EventSourceT::SourceBaseType, EventSourceT>::value);
+        static_assert(AZStd::is_abstract<EventSourceT>::value == false);
+
+        using SourceEventType = typename EventSourceT::SourceEventType;
+        using SourceEventHandlerType = typename EventSourceT::SourceEventHandlerType;
+
+        static void Reflect(AZ::ReflectContext* context)
+        {
+            EventSourceT::Reflect(context);
+
+            if (auto serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+            {
+                serializeContext->Class<EventSourceAdapter<EventSourceT>>()
+                    ->Version(1)
+                    ->Field("Adapted frequency", &EventSourceAdapter<EventSourceT>::m_adaptedFrequency)
+                    ->Field("Event source configuration", &EventSourceAdapter<EventSourceT>::m_eventSource);
+
+                if (auto editContext = serializeContext->GetEditContext())
+                {
+                    editContext
+                        ->Class<EventSourceAdapter<EventSourceT>>(
+                            "Event Source Adapter", "Adapts sensor event source to specified working frequency")
```

Comment:
To discuss: I think that users of sensor components do not need to know about their inner workings (that there are event source adapters and such). This is knowledge for sensor developers.

Reply:
We agreed to remove from editor context everything from event source adapter down. After few thoughts, I decided to leave event source adapter edit context, and stop exposing it in ROS2SensorComponentBase instead. This way, it will not be visible in sensors implementing ROS2SensorComponentBase, but someone using EventSourceAdapter for different purpose will be able to expose its parameters to user.